brainiak/utils/fmrisim.py

```diff
@@ -624,57 +746,148 @@ def double_gamma_hrf(stimfunction,
+            signal_function_temp = signal_function_temp / np.max(
+                signal_function_temp)
+
+        # Add this function to the stack
+        if list_counter == 0:
+            signal_function = np.zeros((len(signal_function_temp), list_num))
+
+        signal_function[:, list_counter] = signal_function_temp
 
     return signal_function
 
 
 def apply_signal(signal_function,
-                 volume_static,
+                 volume_signal,
                  ):
-    """Apply the convolution and stimfunction
+    """Combine the signal volume with its timecourse
 
     Apply the convolution of the HRF and stimulus time course to the
     volume.
 
     Parameters
     ----------
-        signal_function : list, float
-            The one dimensional timecourse of the signal over time.
-            Found by convolving the HRF with the stimulus time course.
 
-        volume_static : multi dimensional array, float
+    signal_function : timepoint by timecourse array, float
+        The timecourse of the signal over time. If there is only one column
+        then the same timecourse is applied to all voxels. If there is more
+        than one column then each column is paired with a non-zero voxel in the
+        volume_signal.
+
+    volume_signal : multi dimensional array, float
+        The volume containing the signal to be convolved
 
 
     Returns
     ----------
     multidimensional array, float
-        Generates the spatial noise volume for these parameters """
+        The convolved signal volume
 
-    # Preset volume
-    signal = np.ndarray([volume_static.shape[0], volume_static.shape[
-        1], volume_static.shape[2], len(signal_function)])
+    """
+
+    # How many timecourses are there within the signal_function
+    timepoints = signal_function.shape[0]
+    timecourses = signal_function.shape[1]
 
-    # Iterate through the brain, multiplying the volume by the HRF
-    for tr_counter in list(range(0, len(signal_function))):
-        signal[0:volume_static.shape[0],
-               0:volume_static.shape[1],
-               0:volume_static.shape[2],
-               tr_counter] = signal_function[tr_counter] * volume_static
+    # Preset volume
+    signal = np.zeros([volume_signal.shape[0], volume_signal.shape[
+        1], volume_signal.shape[2], timepoints])
+
+    # Find all the non-zero voxels in the brain
+    idxs = np.where(volume_signal != 0)
+    if timecourses == 1:
+        # If there is only one time course supplied then duplicate it for
+        # every voxel
+        signal_function = np.matlib.repmat(signal_function, 1, len(idxs[0]))
+
+    elif len(idxs[0]) != timecourses:
+        print('The number of non-zero voxels in the volume and the number of'
+              ' timecourses does not match. Aborting')
+        exit()
+
+    # For each coordinate with a non zero voxel, fill in the timecourse for
+    # that voxel
+    for idx_counter in list(range(0, len(idxs[0]))):
```

Comment:
There are a lot of optimization problems to be solved in this code and I think it would be great in future for people to tackle them. This is certainly one